src/algebra/big_operators.lean

```diff
@@ -307,8 +307,116 @@ lemma prod_eq_one [comm_monoid β] {f : α → β} {s : finset α} (h : ∀x∈s
 calc s.prod f = s.prod (λx, 1) : finset.prod_congr rfl h
   ... = 1 : finset.prod_const_one
 
+@[to_additive finset.sum_filter]
+lemma prod_filter [decidable_eq α] [comm_monoid β]
+  (s : finset α) (f : α → β) (P : α → Prop) [decidable_pred P] :
+s.prod f = (s.filter P).prod f * (s.filter (λ a, ¬ P a)).prod f :=
+by rw [← finset.prod_union (finset.filter_inter_filter_neg_eq s), finset.filter_union_filter_neg_eq]
+
+-- TODO These next few should probably also be proved without the assumption of commutativity.
+@[to_additive finset.Ico_sum_split]
+lemma Ico_prod_split (l n m : ℕ) (h₁ : n ≤ l) (h₂ : l ≤ m) (f : ℕ → β) :
+(Ico n m).prod f = (Ico n l).prod f * (Ico l m).prod f :=
+begin
+  rw prod_filter (Ico n m) f (λ x, x < l),
+  rw filter_not,
+  simp,
+  rw min_eq_right h₂,
+  rw max_eq_right h₁,
+end
+
+-- `to_additive` chokes on this one, so we reprove it below
+lemma Ico_prod_split_first (n m : ℕ) (h : n < m) (f : ℕ → β) :
+(Ico n m).prod f = f n * (Ico (n+1) m).prod f :=
+begin
+  rw Ico_prod_split (n+1),
+  -- and tidy up loose ends:
+  simp,
+  -- sad that `linarith` can't solve these last two
+  exact nat.le_succ _,
+  exact nat.succ_le_of_lt h,
+end
+
+lemma Ico_prod_split_last (n m : ℕ) (h : m > n) (f : ℕ → β) :
+(Ico n m).prod f = (Ico n (m-1)).prod f * f (m-1) :=
+begin
+  rw Ico_prod_split (m-1),
+  rw finset.Ico.pred_singleton,
+  -- and tidy up loose ends:
+  simp,
+  exact nat.lt_of_le_of_lt (nat.zero_le _) h,
+  -- sad that `linarith` can't solve these last two
+  exact nat.pred_le_pred (nat.succ_le_of_lt h),
+  exact nat.sub_le _ _,
+end
+
+local attribute [-simp] add_comm
+local attribute [simp] nat.add_sub_cancel nat.add_sub_cancel_left
+
+@[to_additive finset.Ico_sum_reindex_right]
+lemma Ico_prod_reindex_right (k n m : ℕ) (f : ℕ → β) :
+(Ico n m).prod f = (Ico (n+k) (m+k)).prod (λ x, f (x - k)) :=
```

Comment:
```suggestion
  (Ico n m).prod f = (Ico (n+k) (m+k)).prod (λ x, f (x - k)) :=
```

Reply:
This is a rule which should be focused on `Ico`: `(Ico n m).map (+k) = Ico (k + n) (k + m)`
Especially where `map` knows that `(+k)` is a injective function so it transports nicely through prod. Then `prod_bij` is not necessary.

Reply:
@johoelzl, I've written these lemmas on `Ico` now, but it is not going to help with `reindex_left`, since `\lambda x, x-k` is not injective on nat.

Reply:
Ok, I've worked something out. It's not as pretty as I'd like for `reindex_left`, but much better than before.